R/plot-build.R

```diff
@@ -102,14 +102,9 @@ ggplot_build.ggplot <- function(plot) {
 
   # Train and map non-position scales and guides
   npscales <- scales$non_position_scales()
-  if (npscales$n() > 0) {
-    lapply(data, npscales$train_df)
-    plot$guides <- plot$guides$build(npscales, plot$layers, plot$labels, data)
-    data <- lapply(data, npscales$map_df)
-  } else {
-    # Only keep custom guides if there are no non-position scales
-    plot$guides <- plot$guides$get_custom()
-  }
+  npscales$train(data)
+  plot$guides <- plot$guides$build(npscales, plot$layers, plot$labels, data)
+  data <- npscales$map(data, layers)
```

Comment:
In order to warn about guides when there are no non-position scales, the guide building needed to be executed unconditionally.